### PR TITLE
transformations: stencil bufferization polish

### PR DESCRIPTION
**Pull request by PapyChacal**
Step back on the idea of allocating all apply results to then fold away and instead just fold stores into apply. It's cheaper, simpler, and so far works just as well!

Also rely on the new side-effect value information to improve side-effect analysis.

Both changes allow to bufferize a full open earth compiler example now!

Minor test changes, the bufferization used to blindly allocate buffers for unused results, now it scraps them. I guess it's debatable if that should be its role? Note that it does so by reusing the canonicalization pattern, no logic is actually duplicated.

### CODE REVIEW

File: xdsl/dialects/stencil.py [hunk 1]
```diff
@@ -561,17 +561,20 @@ def parse_operand():
     def get(
         args: Sequence[SSAValue] | Sequence[Operation],
         body: Block | Region,
-        result_types: Sequence[TempType[Attribute]],
+        result_types: Sequence[TempType[Attribute]] | None = None,
+        bounds: StencilBoundsAttr | None = None,
     ):
-        assert len(result_types) > 0
-
+        assert result_types or bounds
```
Comment (superlopuh) [inline xdsl/dialects/stencil.py:567]: What happens if result_types is an empty sequence? Surely you want to have no results?
```suggestion
        assert result_types is not None or bounds is not None
```

File: xdsl/dialects/stencil.py [hunk 2]
```diff
@@ -646,6 +649,14 @@ def get_accesses(self) -> Iterable[AccessPattern]:
                 accesses.append(offsets)
             yield AccessPattern(tuple(accesses))
 
+    def get_bounds(self):
+        if self.bounds is not None:
+            return self.bounds
+        else:
+            assert len(self.res) > 0
```
Comment (superlopuh) [inline xdsl/dialects/stencil.py:656]:
```suggestion
            assert self.res
```

File: xdsl/dialects/stencil.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/stencil.py:567]: Yeah that's semantic!
Basically the op needs a non-empty result list or bounds, so the `(result_types = [], bounds = None)` case is invalid too :slightly_smiling_face:

Comment (superlopuh) [reply xdsl/dialects/stencil.py:567]: Would this work? `result_types: Sequence[TempType[Attribute]] = ()`

Comment (PapyChacal) [reply xdsl/dialects/stencil.py:567]: Sure, but I'd still want to check it for non-emptiness? I'm confused

Comment (superlopuh) [reply xdsl/dialects/stencil.py:567]: This avoids the `None`/ empty ambiguity

Comment (PapyChacal) [reply xdsl/dialects/stencil.py:567]: Done!

**Review by n-io — APPROVED** [review]

**Review by superlopuh — APPROVED** [review]

**Pull request merged**